src/arch/x86/kernel/boot_sys.c

```diff
@@ -291,7 +291,7 @@ parse_mem_map(uint32_t mmap_length, uint32_t mmap_addr)
             printf("\tPhysical Memory Region from %lx size %lx type %d\n", (long)mem_start, (long)mem_length, type);
             if (type == MULTIBOOT_MMAP_USEABLE_TYPE && mem_start >= HIGHMEM_PADDR) {
                 if (!add_mem_p_regs((p_region_t) {
-                mem_start, mem_start + mem_length
+                mem_start, ROUND_UP(mem_start + mem_length, PAGE_BITS),
```

Comment:
Hey,

I'd prefer for this to be a ROUND_DOWN instead of a ROUND_UP. If the E820 map passed through by GRUB has usable memory regions that are a mixture of two memory types (usable and reserved) such that a "usable" memory region ends on a non-page aligned address, and is followed immediately by say, an ACPI reserved region, then that region should be truncated, not rounded up.

If that region happens to only be a single frame in size, then that frame should be excluded from the usable memory that is exported to userspace. Basically the kernel should be extremely pessimistic about which physical frames it exports to userspace as usable memory, and memory which is questionable should be exported as device untypeds.

If I've misunderstood what you're trying to do here, could you either help me with some clarification or else paste me the GRUB memory map (and if you are able to, please also indicate which region is causing the issue).

Reply:
Hi,

After applying the ROUND_DOWN instead of the ROUND_UP we get the following error that we were also seeing before we applied the ROUND_UP:

_utspace_split_alloc@split.c:249 Failed to find any untyped capable of creating an object at address 0xdf6cf000
sel4platsupport_map_paddr@io.c:227 Failed to find a way to map address 0xdf6cfc00
Assertion failed: header != NULL (../projects/util_libs/libplatsupport/src/plat/pc99/acpi/walker.c: acpi_map_table: 51)
seL4 root server abort()ed
Debug halt syscall from user thread 0xffffff829ffd4400 "sel4test-driver"

This error occurs during the init_timer() call within the sel4test-driver main.

The GRUB memory map is as follows:

Parsing GRUB physical memory map
	Physical Memory Region from 0 size 9e400 type 1
	Physical Memory Region from f0000 size 10000 type 2
	Physical Memory Region from 100000 size df5cfc00 type 1
Adding physical memory region 0x100000-0xdf6cf000
	Physical Memory Region from df6cfc00 size 54000 type 4
	Physical Memory Region from df723c00 size 2000 type 3
	Physical Memory Region from df725c00 size 8da400 type 2
	Physical Memory Region from f8000000 size 4000000 type 2
	Physical Memory Region from fed00000 size 400 type 2
	Physical Memory Region from fed20000 size 80000 type 2
	Physical Memory Region from fec00000 size 100000 type 2
	Physical Memory Region from fee00000 size 100000 type 2
	Physical Memory Region from ffb00000 size 500000 type 2
	Physical Memory Region from 100000000 size 180000000 type 1
Adding physical memory region 0x100000000-0x280000000
	Physical Memory Region from 280000000 size 20000000 type 1
Adding physical memory region 0x280000000-0x2a0000000

Reply:
@Apave24 Sorry for the delay, I'm looking into this and trying to understand what's going on -- thanks for your memory map. I may reply asking for you to add a printf to the your kernel source so I can gain a bit more information about what is happening. Thanks for your patience~

Reply:
@Apave24 can you please provide the full output from `debug_print_bootinfo` so I can see the untyped objects that are being created by seL4?

Reply:
@pingerino here is the output:
Node 0 of 1
IOPT levels:     4
IPC buffer:      0x7a5000
Empty slots:     [1066 --> 4096)
sharedFrames:    [0 --> 0)
userImageFrames: [17 --> 950)
userImagePaging: [12 --> 16)
untypeds:        [950 --> 1066)
Initial thread domain: 0
Initial thread cnode size: 12
List of untypeds
------------------
Paddr    | Size   | Device
0x100000 | 20 | 0
0x200000 | 21 | 0
0x400000 | 22 | 0
0xfb5000 | 12 | 0
0xfb6000 | 13 | 0
0xfb8000 | 15 | 0
0xfc0000 | 18 | 0
0x1000000 | 24 | 0
0x2000000 | 25 | 0
0x4000000 | 26 | 0
0x8000000 | 27 | 0
0x10000000 | 28 | 0
0x20000000 | 29 | 0
0x40000000 | 30 | 0
0x80000000 | 30 | 0
0xc0000000 | 28 | 0
0xd0000000 | 27 | 0
0xd8000000 | 26 | 0
0xdc000000 | 25 | 0
0xde000000 | 24 | 0
0xdf000000 | 22 | 0
0xdf400000 | 21 | 0
0xdf600000 | 19 | 0
0xdf680000 | 18 | 0
0xdf6c0000 | 15 | 0
0xdf6c8000 | 14 | 0
0xdf6cc000 | 13 | 0
0xdf6ce000 | 12 | 0
0x100000000 | 32 | 0
0x200000000 | 31 | 0
0x280000000 | 28 | 0
0x290000000 | 27 | 0
0x298000000 | 26 | 0
0x29c000000 | 25 | 0
0x29e000000 | 24 | 0
0x29f000000 | 23 | 0
0x29f800000 | 22 | 0
0x29fc00000 | 21 | 0
0x29fe00000 | 19 | 0
0x29fe80000 | 18 | 0
0x29fec0000 | 15 | 0
0x29fec8000 | 13 | 0
0x29feca000 | 12 | 0
0 | 20 | 1
0xdf6cf000 | 12 | 1
0xdf6d0000 | 16 | 1
0xdf6e0000 | 17 | 1
0xdf700000 | 20 | 1
0xdf800000 | 23 | 1
0xe0000000 | 28 | 1
0xf0000000 | 27 | 1
0xf8000000 | 26 | 1
0xfc000000 | 25 | 1
0xfe000000 | 23 | 1
0xfe800000 | 22 | 1
0xfec01000 | 12 | 1
0xfec02000 | 13 | 1
0xfec04000 | 14 | 1
0xfec08000 | 15 | 1
0xfec10000 | 16 | 1
0xfec20000 | 17 | 1
0xfec40000 | 18 | 1
0xfec80000 | 19 | 1
0xfed00000 | 19 | 1
0xfed80000 | 18 | 1
0xfedc1000 | 12 | 1
0xfedc2000 | 13 | 1
0xfedc4000 | 14 | 1
0xfedc8000 | 15 | 1
0xfedd0000 | 16 | 1
0xfede0000 | 17 | 1
0xfee01000 | 12 | 1
0xfee02000 | 13 | 1
0xfee04000 | 14 | 1
0xfee08000 | 15 | 1
0xfee10000 | 16 | 1
0xfee20000 | 17 | 1
0xfee40000 | 18 | 1
0xfee80000 | 19 | 1
0xfef00000 | 20 | 1
0xff000000 | 23 | 1
0xff800000 | 22 | 1
0xffc00000 | 21 | 1
0xffe00000 | 20 | 1
0xfff00000 | 19 | 1
0xfff80000 | 18 | 1
0xfffc0000 | 17 | 1
0xfffe0000 | 16 | 1
0xffff0000 | 15 | 1
0xffff8000 | 14 | 1
0xffffc000 | 13 | 1
0xffffe000 | 12 | 1
0xfffff000 | 11 | 1
0xfffff800 | 10 | 1
0xfffffc00 | 9 | 1
0xfffffe00 | 8 | 1
0xffffff00 | 7 | 1
0xffffff80 | 6 | 1
0xffffffc0 | 5 | 1
0xffffffe0 | 4 | 1
0x2a0000000 | 29 | 1
0x2c0000000 | 30 | 1
0x300000000 | 32 | 1
0x400000000 | 34 | 1
0x800000000 | 35 | 1
0x1000000000 | 36 | 1
0x2000000000 | 37 | 1
0x4000000000 | 38 | 1
0x8000000000 | 46 | 1
0x408000000000 | 45 | 1
0x608000000000 | 44 | 1
0x708000000000 | 43 | 1
0x788000000000 | 42 | 1
0x7c8000000000 | 41 | 1
0x7e8000000000 | 40 | 1
0x7f8000000000 | 39 | 1
Untyped summary
1 untypeds of size 4
1 untypeds of size 5
1 untypeds of size 6
1 untypeds of size 7
1 untypeds of size 8
1 untypeds of size 9
1 untypeds of size 10
1 untypeds of size 11
8 untypeds of size 12
7 untypeds of size 13
5 untypeds of size 14
7 untypeds of size 15
5 untypeds of size 16
5 untypeds of size 17
7 untypeds of size 18
6 untypeds of size 19
5 untypeds of size 20
4 untypeds of size 21
5 untypeds of size 22
4 untypeds of size 23
3 untypeds of size 24
4 untypeds of size 25
4 untypeds of size 26
4 untypeds of size 27
4 untypeds of size 28
2 untypeds of size 29
3 untypeds of size 30
1 untypeds of size 31
2 untypeds of size 32
1 untypeds of size 34
1 untypeds of size 35
1 untypeds of size 36
1 untypeds of size 37
1 untypeds of size 38
1 untypeds of size 39
1 untypeds of size 40
1 untypeds of size 41
1 untypeds of size 42
1 untypeds of size 43
1 untypeds of size 44
1 untypeds of size 45
1 untypeds of size 46

Reply:
Hi @Apave24, do you remember if this output is from an unmodified kernel or one with ROUND_UP() applied? Do you still have access to the hardware that was triggering this issue?

Reply:
I think the ROUND_DOWN approach is required, the user level error is due to the allocator not looking in the device ut pool for the device untyped.